Add type hints for p2p package and enable stricter checks

cburgdorf opened the pull request:
### What was wrong?

P2P was missing many type hints and didn't have the `--disallow-untyped-defs` check yet which means that new code could potentially land without type hints (making the fight for 100 % coverage a bit of a fight against windmills)

### How was it fixed?

Added type hints and enabled `--disallow-untyped-defs` check

#### Cute Animal Picture

![put a cute animal picture link inside the parentheses](https://metrouk2.files.wordpress.com/2015/07/fuzzberta-the-guinea-pig-dressed-as-people-and-things-princess-leia-e1436630730651.jpg)

gsalgado commented on p2p/chain.py:
```diff
@@ -403,7 +404,9 @@ def _handle_block_headers(self, headers: List[BlockHeader]) -> None:
         downloaded: List[DownloadedBlockPart] = []
-        for (receipts, (receipt_root, trie_dict_data)) in zip(receipts_by_block, receipts_tries):
+        # TODO: figure out why mypy is losing the type of the receipts_tries
+        # so we can get rid of the ignore
```
I'm not sure what this means. Shouldn't mypy treat receipts_tries's type as Any since that's what self.wait() returns?

cburgdorf replied:
[`wait(...)`](https://github.com/cburgdorf/py-evm/blob/69892c3ff720ad437d76ca29d1c94df81c5380e1/p2p/service.py#L45-L52) is generic now and should return the right type which works in principle but fails at this place.

gsalgado commented on p2p/discovery.py:
```diff
@@ -125,25 +128,23 @@ def _get_handler(self, cmd: Command) -> Callable[[kademlia.Node, List[Any], byte
         return self._max_neighbours_per_packet_cache
 
-    def connection_made(self, transport):
+    def connection_made(self, transport: Any) -> None:
```
Can't we use an actual transport type here?

cburgdorf replied:
Ah, good hint. It's one of these things that I skipped and never came back to. So this represents something like a `asyncio.DatagramTransport` right?

cburgdorf replied:
Fixed by https://github.com/ethereum/py-evm/pull/947/commits/87d585801d1da8b99dbe60595299dab5734c671f

gsalgado commented on p2p/eth.py:
```diff
@@ -108,7 +108,7 @@ class ETHProtocol(Protocol):
     logger = logging.getLogger("p2p.eth.ETHProtocol")
 
-    def send_handshake(self, head_info):
+    def send_handshake(self, head_info: Any) -> None:
```
Why use Any here?

cburgdorf replied:
Fixed by https://github.com/ethereum/py-evm/pull/947/commits/7ba3b72ffe102aec18bf3095159841326ac79afe

gsalgado commented on p2p/les.py:
```diff
@@ -86,7 +95,7 @@ def decode_payload(self, rlp_data: bytes) -> Generator[Tuple[str, Any], None, No
                 yield key, value
 
-    def encode_payload(self, data):
+    def encode_payload(self, data: Dict[str, bytes]) -> bytes:  # type: ignore
```
Why do we need the `type: ignore` comment here?

cburgdorf replied:
It doesn't conform to the signature of the super type.

We could also do this:
```
    def encode_payload(self, data: Union[_DecodedMsgType, sedes.CountableList]) -> bytes:
        response = [
            (key, self.items_sedes[key].serialize(value))
            for key, value
            in sorted(data.items())  # type: ignore
        ]
        return super().encode_payload(response)
```
So, we conform to the interface but we then ignore that `data.items()` is actually invalid because `items()` doesn't exist on a Tuple.

gsalgado commented on p2p/les.py:
```diff
@@ -254,7 +263,7 @@ class LESProtocol(Protocol):
     cmd_length = 15
 
-    def send_handshake(self, head_info):
+    def send_handshake(self, head_info: Any) -> None:
```
Should be able to use `HeadInfo` here, no?

cburgdorf replied:
Ah, good point. I skipped that because I didn't notice `HeadInfo` actually exists.

cburgdorf replied:
Ah, turns out it's actually `ChainInfo` there (because `genesis_hash` doesn't exist on `HeadInfo`) 

https://github.com/ethereum/py-evm/pull/947/commits/69892c3ff720ad437d76ca29d1c94df81c5380e1

Notice, I'll rebase these changes into the right place once this passed a complete review.

gsalgado commented on p2p/les.py:
```diff
@@ -354,7 +363,7 @@ class LESProtocolV2(LESProtocol):
     cmd_length = 21
 
-    def send_handshake(self, head_info):
+    def send_handshake(self, head_info: Any) -> None:
```
And here

cburgdorf replied:
Both places fixed.

gsalgado commented on p2p/lightchain.py:
```diff
@@ -257,14 +260,14 @@ def register_peer(self, peer: BasePeer) -> None:
         self.logger.info("Stopping LightPeerChain...")
 
-    async def _wait_for_reply(self, request_id: int) -> Dict[str, Any]:
+    async def _wait_for_reply(self, request_id: int) -> Any:
```
Why this change?

cburgdorf replied:
Good point, fixed and explained here https://github.com/ethereum/py-evm/pull/947/commits/6cf408c432c74f78d2029ced398774a6c85f142d

gsalgado commented on p2p/peer.py:
```diff
@@ -238,11 +239,12 @@ def remove_subscriber(self, subscriber: 'asyncio.Queue[PEER_MSG_TYPE]') -> None:
 
         Raises HandshakeFailure if the handshake is not successful.
         """
+        self.base_protocol = cast(P2PProtocol, self.base_protocol)
```
`BasePeer.__init__()` does `self.base_protocol = P2PProtocol(self)`, so I'm wondering why this is needed

cburgdorf replied:
Yep, we can do better here. Checkout https://github.com/ethereum/py-evm/pull/947/commits/a8632048824259f6816b57aca92c7c72d8c36d24

Basically we:

- get rid of the type of `self.base_protocol` on the class level so that it is really only defined by `__init__`
- return early in `get_protocol_command_for` so that the intermediate variable doesn't have to be types as `Union[Protocol, P2PProtocol]`

gsalgado commented on p2p/peer.py:
```diff
@@ -258,11 +260,12 @@ def remove_subscriber(self, subscriber: 'asyncio.Queue[PEER_MSG_TYPE]') -> None:
 
         Raises HandshakeFailure if the handshake is not successful.
         """
+        self.base_protocol = cast(P2PProtocol, self.base_protocol)
```
Ditto

cburgdorf replied:
solved, see previous comment

gsalgado commented on p2p/peer.py:
```diff
@@ -347,7 +350,7 @@ def close(self):
                 return
 
-    async def read_msg(self) -> Tuple[protocol.Command, protocol._DecodedMsgType]:
+    async def read_msg(self) -> Tuple[protocol.Command, Dict[str, Any]]:
```
Why this change?  This method can return things other than a dict, that's why it used `protocol._DecodedMsgType`

cburgdorf replied:
True, some temporary change that made it through my own review.

gsalgado replied:
Are you going to revert it?

cburgdorf replied:
Sorry, fixed by https://github.com/ethereum/py-evm/pull/947/commits/1bbab42b47e15f6453894dcc993e1d9bab925a05

gsalgado commented on p2p/peer.py:
```diff
@@ -361,12 +364,13 @@ def close(self):
 
     def handle_p2p_msg(self, cmd: protocol.Command, msg: protocol._DecodedMsgType) -> None:
         """Handle the base protocol (P2P) messages."""
+        self.base_protocol = cast(P2PProtocol, self.base_protocol)
```
Another cast I wouldn't think to be needed.

cburgdorf replied:
solved, see previous comment

gsalgado commented on p2p/peer.py:
```diff
@@ -478,6 +482,7 @@ def disconnect(self, reason: DisconnectReason) -> None:
 
         :param reason: An item from the DisconnectReason enum.
         """
+        self.base_protocol = cast(P2PProtocol, self.base_protocol)
```
And here too

cburgdorf replied:
solved, see previous comment

cburgdorf commented:
@gsalgado Thanks for the sharp review. I was able to tighten types for a lot of places. I think I addressed all your comments. Once you approve, I'll make sure to move all the wip commits into the first one before merging.

That said, CircleCI seems to have hiccups right now so this won't go green until Circle runs smoothly again.

cburgdorf commented:
Oh, just noticed its not CircleCI that has a hiccup, it's actually blocked by https://github.com/ethereum/py-evm/issues/953
